03-path-application-development/309-deploying-a-chart-repository/readme.adoc

```diff
@@ -0,0 +1,264 @@
+
+== ChartMuseum
+
+Chart Museum is an Open Source Chart Repository, it helps you to deploy your
+custom charts to an Amazon S3 bucket and allows you to easily access your charts
+like you would a `stable` release from something like 
+link:https://hub.kubeapps.com/[Kubeapps]. 
+
+== Installing ChartMuseum
+
+To install `chartmuseum` you first need to add the incubator helm repo.
+
+    $ helm repo add incubator https://kubernetes-charts-incubator.storage.googleapis.com
+    "incubator" has been added to your repositories
+
+After you have added the incubator repo, you can install like any other `helm`
+package.
+
+    helm install incubator/chartmuseum --name chartmuseum
+
+This will deploy the `chartmuseum` using the standard configuration without
+long term storage for your charts.
+
+Once you have this installed you can use `kubectl port-forward` to get access to
+the service.
+
+    $ kubectl port-forward $(kubectl get po -l app=chartmuseum -o jsonpath="{.items..metadata.name}") 8080:8080
+    Forwarding from 127.0.0.1:8080 -> 8080
+    Forwarding from [::1]:8080 -> 8080
+
+== Configuration
+
+Now that you have `chartmuseum` running in your cluster we need to make a couple
+edits, first we should use the built in mechanisms for storage, exposing, and 
+authentication.
+
+=== Remote Storage
+
+With the default deploy the charts are stored on a mounted volume, for a
+Kubernetes deployment where you don't always run on the same node this can cause
+issues where the server can't find the charts. To fix this we need to create an
+Amazon S3 Bucket for our charts.
+
+    $ export CHART_BUCKET=s3://example-chart-store-$(cat /dev/random | LC_ALL=C tr -dc "[:alpha:]" | tr '[:upper:]' '[:lower:]' | head -c 32)
+    $ aws s3 mb $CHART_BUCKET
+    make_bucket: example-chart-store-zohaenqynbcctothqpmacoicaytzjach
+
+Now that the bucket has been created we need to edit the node IAM policy to
+allow access to read, write, delete.
+
+    kops edit cluster example.cluster.k8s.local
+
+You will need to add `additionalPolicies` that will allow all nodes to have
+access to the bucket, to do so use:
+
+    additionalPolicies:
+      node: |
+        [
+          {
+            "Sid": "AllowListObjects",
+            "Effect": "Allow",
+            "Action": [
+              "s3:ListBucket"
+            ],
+            "Resource": "arn:aws:s3:::{BUCKET_NAME}"
+          },
+          {
+            "Sid": "AllowObjectsCRUD",
+            "Effect": "Allow",
+            "Action": [
+              "s3:DeleteObject",
+              "s3:GetObject",
+              "s3:PutObject"
+            ],
+            "Resource": "arn:aws:s3:::{BUCKET_NAME}/*"
+          }
+        ]
+
+WARNING: Make sure to replace `{BUCKET_NAME}` in each of the statements with the
+`$CHART_BUCKET` without `s3://`
+
+After you have saved the file you can update the cluster thus adding the new
+policies to the nodes.
+
+    kops update cluster example.cluster.k8s.local --yes
```

Comment:
IIRC it might take some time for this command to return. Add a note for that?

Reply:
This one doesn't it's the rolling-update that takes time. This just applies the policy to the instance role.